File: medbrains/apps/ios/Packages/MedBrainsKit/Sources/MedBrainsKit/ApiClient.swift
import Foundation

/// The wire contract every MedBrains device app keeps. Mirrors
/// apps/mobile-staff/src/api/client.ts: bearer token from the secret store,
/// `X-MedBrains-Client: mobile-<variant>` on every request (which is what
/// makes login answer body tokens instead of cookies — a native client on
/// Android is refused a Secure cookie over plain HTTP), JSON in and out,
/// and a 401 anywhere signs the session out.
public struct ApiError: Error, LocalizedError, Sendable {
    public let status: Int
    public let message: String

    public var errorDescription: String? { message }
}

public enum HttpMethod: String, Sendable {
    case get = "GET", post = "POST", put = "PUT", patch = "PATCH", delete = "DELETE"
}

public actor ApiClient {
    public let baseURL: URL
    public let clientName: String
    private let secrets: SecretStore
    private let session: URLSession
    private var onUnauthorized: (@Sendable () async -> Void)?

    /// - Parameter variant: `staff`, `patient`, `camp`, `vendor` — becomes `mobile-<variant>`.
    public init(baseURL: URL, variant: String, secrets: SecretStore, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.clientName = "mobile-\(variant)"
        self.secrets = secrets
        self.session = session
    }

    public func setUnauthorizedHandler(_ handler: @escaping @Sendable () async -> Void) {
        onUnauthorized = handler
    }

    /// Request and decode. `Empty` for routes that answer nothing useful.
    public func request<T: Decodable>(_ method: HttpMethod, _ path: String, body: (some Encodable)? = Optional<Empty>.none, as type: T.Type = T.self) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path.hasPrefix("/") ? String(path.dropFirst()) : path))
        req.httpMethod = method.rawValue
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(clientName, forHTTPHeaderField: "X-MedBrains-Client")
        if let jwt = try? secrets.read(.jwt) {
            req.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder.medbrains.encode(body)
        }
        req.timeoutInterval = 20
        let (data, response) = try await session.data(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            if status == 401 { await onUnauthorized?() }
            throw ApiError(status: status, message: Self.errorMessage(from: data, status: status))
        }
        if T.self == Empty.self { return Empty() as! T }
        return try JSONDecoder.medbrains.decode(T.self, from: data)
    }

    private static func errorMessage(from data: Data, status: Int) -> String {
        if let payload = try? JSONDecoder().decode(ErrorPayload.self, from: data) {
            return payload.detail ?? payload.error
        }
        return HTTPURLResponse.localizedString(forStatusCode: status)
    }
}

public struct Empty: Codable, Sendable {
    public init() {}
}

struct ErrorPayload: Decodable {
    let error: String
    let detail: String?
}

public extension JSONDecoder {
    /// The server speaks snake_case and RFC 3339 dates.
    static let medbrains: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .iso8601
        return d
    }()
}

public extension JSONEncoder {
    static let medbrains: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        e.dateEncodingStrategy = .iso8601
        return e
    }()
}

File: medbrains/apps/android/kit/src/main/kotlin/com/medbrains/kit/ApiClient.kt
package com.medbrains.kit

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * The wire contract every MedBrains device app keeps. Mirrors
 * apps/mobile-staff/src/api/client.ts: bearer token from the secret store,
 * `X-MedBrains-Client: mobile-<variant>` on every request (which is what makes
 * login answer body tokens instead of cookies — a native client on Android is
 * refused a Secure cookie over plain HTTP), JSON in and out, and a 401
 * anywhere signs the session out.
 */
class ApiError(val status: Int, message: String) : Exception(message)

class ApiClient(
    val baseUrl: String,
    variant: String,
    private val secrets: SecretStore,
    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .build(),
) {
    val clientName = "mobile-$variant"
    var onUnauthorized: (suspend () -> Unit)? = null

    val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        coerceInputValues = true
    }

    /** Raw call; the typed helpers below decode with kotlinx.serialization. */
    suspend fun call(method: String, path: String, body: String? = null): String = withContext(Dispatchers.IO) {
        val builder = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .header("Accept", "application/json")
            .header("X-MedBrains-Client", clientName)
        secrets.read(SecretKey.JWT)?.let { builder.header("Authorization", "Bearer $it") }
        val requestBody = body?.toRequestBody(JSON_MEDIA)
        builder.method(method, if (method == "GET" || method == "DELETE" && body == null) null else requestBody ?: "".toRequestBody(JSON_MEDIA))
        http.newCall(builder.build()).execute().use { response ->
            val text = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                if (response.code == 401) onUnauthorized?.invoke()
                throw ApiError(response.code, errorMessage(text, response.code))
            }
            text
        }
    }

    suspend inline fun <reified T> get(path: String): T = json.decodeFromString(call("GET", path))

    suspend inline fun <reified B, reified T> post(path: String, body: B): T =
        json.decodeFromString(call("POST", path, json.encodeToString(body)))

    suspend inline fun <reified T> post(path: String): T = json.decodeFromString(call("POST", path, "{}"))

    suspend inline fun <reified B, reified T> put(path: String, body: B): T =
        json.decodeFromString(call("PUT", path, json.encodeToString(body)))

    suspend inline fun <reified T> put(path: String): T = json.decodeFromString(call("PUT", path, "{}"))

    suspend inline fun <reified B, reified T> patch(path: String, body: B): T =
        json.decodeFromString(call("PATCH", path, json.encodeToString(body)))

    private fun errorMessage(text: String, status: Int): String = runCatching {
        json.decodeFromString<ErrorPayload>(text).let { it.detail ?: it.error }
    }.getOrDefault("HTTP $status")

    private companion object {
        val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
    }
}

@kotlinx.serialization.Serializable
internal data class ErrorPayload(val error: String, val detail: String? = null)
